153_save_sensor_data/sensor_data_store.py

```diff
@@ -0,0 +1,238 @@
+# Seting altitude is disregarded when ambient pressure is given to the system.
+# Ambient Pressure, set with scd.ambient_pressure = PRESSURE_VALUE. Default
+# value is 1013.25 mBar   uint16 700 to 1400 range, 0 sets to default.
+def get_altitude():
+    if len(sys.argv) > 1:
+        try:
+            altitude = int(sys.argv[1])
+            print(f"Altitude entered: {altitude} m")
+            scd.altitude = altitude
+        except ValueError:
+            print("Parameter must be an int for altitude, "
+                  "defaulting to 1013.25 mBar")
+    else:
+        print("No altitude entered. Defaulting to 1013.25 mBar")
+
+def get_interval_data(time_elapsed):
+    cO2_ppm = scd.CO2
+    temperature = scd.temperature #in *C
+    humidity = scd.relative_humidity
+    interval_data = dict([('seconds', time_elapsed), ('co2', cO2_ppm),
+                         ('temp',temperature), ('humidity',humidity)])
+    return interval_data
+
+def sensor_loop():
+    WRITE_QTY = 30 # How many interpolated time steps are desired
+    desired_time_step = 1 # For interpolated data, in seconds
+    interpolated_time = 0
+    debugLiveData = True
+    debugInterpolatedData = False
+    # If the sensor does not respond to 50 requests a runtime error is thrown,
+    # but usually can be ignored. If it happens more than a few times over
+    # the span of a few seconds, the sensor needs to be reset.
+    error_count = 0
+    start_time = time.time()
+    sensor_data = []
+    interpolated_data = []
+    while True:
+        current_time = time.time()
+        time_elapsed = current_time - start_time
+        # Check for new data, available every 2 seconds
+        try:
+            if scd.data_available: # If fresh data is available, get it
+                error_count = 0
+                interval_data = get_interval_data(time_elapsed)
+                sensor_data.append(interval_data)
+                if debugLiveData: # print raw data to screen to debug sensor
+                    print(f"Time: {interval_data['seconds']:<10.2f} "
+                          f"CO2: {interval_data['co2']:>6.1f} ppm    "
+                          f"T: {interval_data['temp']:<3.2f}°C    "
+                          f"Humidity: {interval_data['humidity']:<3.2f}%")
+        except RuntimeError as e:
+            # Occasionally sensor does not want to respond
+            # Ordinarily this should be I2C read error: max entries reached
+            print(e) # Print the error
+            error_count += 1
+            if error_count > 10:
+                print("SENSOR FAILURE... killing program")
+                sys.exit() # Kill the program because something bad is happening
+
+        # Check if there is data to interpolate
+        while (len(sensor_data) > 1 and
+           sensor_data[-1]['seconds'] > interpolated_time + desired_time_step):
+            # Get the smallest pair without going under
+            for index, data_set in enumerate(reversed(sensor_data)):
```

Comment:
Index is actually used, but not within the for loop. It's just counting how far it goes through the data so that number can be used here:

            largest_set_without_going_over = sensor_data[-(index + 1)]